crates/block-producer/src/server.rs

```diff
@@ -30,153 +28,113 @@ use crate::{
-/// components.
-pub struct BlockProducer {
-    batch_builder: BatchBuilder,
-    block_builder: BlockBuilder,
-    batch_budget: BatchBudget,
-    block_budget: BlockBudget,
-    state_retention: usize,
-    expiration_slack: u32,
-    rpc_listener: TcpListener,
-    store: StoreClient,
-    chain_tip: BlockNumber,
+/// Block producer's configuration.
+pub struct BlockProducerConfig {
+    /// The address of the block producer component.
+    pub block_producer_address: SocketAddr,
+    /// The address of the store component.
+    pub store_address: SocketAddr,
+    /// The address of the batch prover component.
+    pub batch_prover: Option<Url>,
+    /// The address of the block prover component.
+    pub block_prover: Option<Url>,
+    /// The interval at which to produce batches.
+    pub batch_interval: Duration,
+    /// The interval at which to produce blocks.
+    pub block_interval: Duration,
 }
 
-impl BlockProducer {
-    /// Performs all setup tasks required before [`serve`](Self::serve) can be called.
-    ///
-    /// This includes connecting to the store and retrieving the latest chain state.
-    pub async fn init(
-        listener: TcpListener,
-        store_address: SocketAddr,
-        batch_prover: Option<Url>,
-        block_prover: Option<Url>,
-        batch_interval: Duration,
-        block_interval: Duration,
-    ) -> Result<Self> {
-        info!(target: COMPONENT, endpoint=?listener, store=%store_address, "Initializing server");
-
-        let store_url = format!("http://{store_address}");
-        let channel = tonic::transport::Endpoint::try_from(store_url.clone())
-            .with_context(|| format!("failed to create store endpoint for {store_url}"))?
-            .connect()
-            .await
-            .with_context(|| format!("failed to connect to store on {store_url}"))?;
-
-        let store = store_client::ApiClient::with_interceptor(channel, OtelInterceptor);
-        let store = StoreClient::new(store);
-
-        let latest_header = store.latest_header().await.context("failed to get latest header")?;
-        let chain_tip = latest_header.block_num();
-
-        info!(target: COMPONENT, "Server initialized");
-
-        Ok(Self {
-            block_builder: BlockBuilder::new(store.clone(), block_prover, block_interval),
-            batch_builder: BatchBuilder::new(
-                store.clone(),
-                SERVER_NUM_BATCH_BUILDERS,
-                batch_prover,
-                batch_interval,
-            ),
-            batch_budget: BatchBudget::default(),
-            block_budget: BlockBudget::default(),
-            state_retention: SERVER_MEMPOOL_STATE_RETENTION,
-            expiration_slack: SERVER_MEMPOOL_EXPIRATION_SLACK,
-            store,
-            rpc_listener: listener,
-            chain_tip,
+/// Serves the block-producer RPC API, the batch-builder and the block-builder.
+///
+/// Note: this blocks until one of the servers die.
```

Comment:
Nit: blocking means something specific and slightly different, i.e. blocking calls should never be run on async.

```suggestion
/// Note: Executes in place (i.e. not spawned) and will run indefinitely until
///       a fatal error is encountered.
```